### ISSUE TITLE
Why must background acrylic be discontinued in WinUI 3? That will surely devastate and disappoint many, for obvious reasons

### ISSUE THREAD

**Issue by carmellolb** [microsoft/microsoft-ui-xaml#2236]
Background acrylic is a key feature of Fluent Design, and without it, it will never be the same.

I feel there must be a deeper reason than performance alone, because there must be a way to improve performance, it seems so unlike the WinUI team to simply give up on something so huge like this.

Simply go around to people (family, friends, coworkers, etc) that use Windows 10, and ask them this question: "Do you know the acrylic (blur like) texture on the sidebars and different other areas of apps on your computer?"
I'll guarantee you that they will respond with something along the lines of "yes, I do. It came around a while ago and it is so nice and is a great addition!" In fact, someone I know who doesn't even know much about technology at all commented "Yeah, I do. It kind of has become part of the whole brand of the system, it wouldn't be the same without it." This person doesn't even know much about it, she just likes it. This along with many other incidents are a clear example of the impact it has on users.

Also, it can have an impact on users without them even realizing it's there, but using the software. When you then remove it, they will feel like something is missing, because it provides depth and hierarchy and without that, they will be confused.

It has BECOME part of the brand of Fluent Design. You can't just remove it, it would be like removing all of the red from YT, or removing the blue from FB. It's not a color, it's a texture, but it has a similar impact.

Also, software that previously was stunning and gorgeous and showed aesthetic will now be a plain, bland, grey surface with no texture at all. And wasn't one of the components of Fluent "Material"? Plain grey is not "material". There is no texture at all.

BG Acrylic was brilliant, because it provided depth and was also gorgeous. Everyone I know loves it. There is a similar thing on macOS and iOS, without many components though, simply just the blur. Acrylic is so much nicer. but don't you think that if there is something like it on those systems and it works just fine and has for years, that it must be possible on Windows as well?

What is really going on here? Why is it being discontinued? Do you realize the impact that this change will have on thousands, maybe even millions of users?

In my opinion, this is a TERRIBLE thing to do. First, you removed it from NavigationView. It's still not on TabView. And now you plan to simply remove it altogether, rendering it impossible for people to even have this key feature placed anywhere? Why? Just why?

I surely hope that a way to improve it and not cancel it will be devised before WinUI 3, or at least in time for a release shortly after. Because I don't think removing it is the right choice. In fact I think it's the worst thing that could be done.

P.S. - I use apps with background acrylic on a daily basis and not a single one of them as any performance issues.

**Comment by mdtauk** [comment]
I believe it is just temporarily unsupported - whilst more of the code which handles Acrylic is removed from the OS and coded anew in the open source project.

**Comment by carmellolb** [comment]
Really? What about this then? https://twitter.com/WithinRafael/status/1230192945355096064

**Comment by chingucoding** [comment]
While I am not sure what document he is referring to in the screenshot, @ryandemopoulos said this: 
https://twitter.com/WithinRafael/status/1230197954280951808?s=20

**Comment by mdtauk** [comment]
![image](https://user-images.githubusercontent.com/7389110/78512083-931bdd80-7799-11ea-916c-492d2f0ab266.png)

I believe the OS provides the backdrop which the Acrylic Brush uses as its underlying layer, which it then adds the noise, tint, and luminosity layers on top of.  The In-App Acrylic is given it's underlying layer by the app renderer.

The HostBackdrop can't be done outside of the OS providing the layer - and so finding a new way to do that is the technical challenge being mentioned.

This will probably require some re-engineering of the OS, and so would also need to be backported to all the OS versions that WinUI 3 will support.

This may not actually happen, and instead the WinUI team may need to write some low level code, which can run in userspace, to replicate the same features.

I'm not sure how they could do it that second way, but I am certainly not a proper programmer, nor do I know the Windows source code :)  Only that it probably requires UXTheme access, which is a protected OS component.

I certainly hope we get it for WinUI 3.1 onwards - because to lose that visual material would be devastating to Fluent design, and it seems like Windows 10X has found a way to do it / or the applications are calling into the Windows.Xaml namespace, which apparently wont be possible from WinUI 3.x apps made by others.

**Comment by chingucoding** [comment]
If I recall correctly WPF apps can have transparent backdrop. Since the actual calculation of an AcrylicBrush is already (partially) contained in WinUI 2, the main problem is probably "only" supporting this transparent backdrop. But since WPF apps are able to do it, there should already exist some kind of interface for that. 

How this will be implemented however with WinUI 3 is nothing I can comment on, so take my comment with a grain of salt.

**Comment by mdtauk** [comment]
> If I recall correctly WPF apps can have transparent backdrop. Since the actual calculation of an AcrylicBrush is already (partially) contained in WinUI 2, the main problem is probably "only" supporting this transparent backdrop. But since WPF apps are able to do it, there should already exist some kind of interface for that.
> 
> How this will be implemented however with WinUI 3 is nothing I can comment on, so take my comment with a grain of salt.

If the Composition layer could be given the screen "framebuffer" minus the app window itself, then sure a new Acrylic Brush could be implemented.  WPF runs with a more permissive system by default, and so I suspect it is the UWP side of the equation, which will have the hardest time.

WinUI Desktop may have some work-arounds which could allow it.

Again, let's hope it is not too far behind WinUI 3.0

**Comment by weltkante** [comment]
As far as I remember WPF transparent windows also aren't native, they are hacked around UpdateLayeredWindow, not using direct rendering but transferring a snapshot through a GDI surface. It will probably cause lots of technical problems for WinUI when you try to mix this hack with the composition API and animations.

**Comment by shaheedmalik** [comment]
In my opinion, WinUI shouldn't be released until it right. Something like Acrylic Brush is vital on creating consistency.

**Comment by carmellolb** [comment]
I agree with this.

**Comment by niels9001** [comment]
Would be great to get some clarity on this. It's one of the key visual elements that make Fluent apps recognizable (and visually very pleasant to look at :)). Additionally, it's one of the most common visual effects in Windows with default controls supporting it out of the box.

This feature should be in 3.0, or at least 3.1!

**Comment by Felix-Dev** [comment]
@niels9001 The team has said HostBackdrop acrylic won't be in WinUI 3.0 because of technical challenges arising by lifting the WinUI components out of the OS stack. They also said they want to get this working in the future. I don't think they can currently give any new update on this. 

In the meantime, I would suggest to get the upvotes on this issue (or a new issue meant for tracking) started so that by the time WinUI 3.0 is out, the team can see this is a highly requested feature from the community, making it more probable that support will be added sooner rather than later post WinUI 3.0 releaase.

**Comment by shaheedmalik** [comment]
A faux one like @mdtauk posted should be used in the interim should be used then.

**Comment by carmellolb** [comment]
I agree.

**Comment by carmellolb** [comment]
> > If it can't be done correctly, make a faux version of it that emulates it until it is done correctly, and then when it is updated it will still show.
> > That would be much better than not having it at all.
> 
> I have suggested this in the past by taking the Desktop Wallpaper/Colour and using that as a base layer. I thought it would be good as a fallback for Tablet Mode, Mobile, and for low powered devices.
> 
> [Inactive Acrylic effect (CreateHostBackDropBrush) - alternative options/fallback](https://aka.ms/AA81xym)
> 
> **Example**
> 
> ![image](https://user-images.githubusercontent.com/7389110/78603538-38918880-7850-11ea-9757-4357c6e092f6.png)
> _Current Acrylic_
> 
> ![image](https://user-images.githubusercontent.com/7389110/78603472-1c8de700-7850-11ea-979a-8a5e9813efd4.png)
> _Faux Alternative Acrylic_

This looks great - except it appears to be missing the noise layer.

**Comment by mdtauk** [comment]
> This looks great - except it appears to be missing the noise layer - correct me if I'm wrong.
>

![image](https://user-images.githubusercontent.com/7389110/78609864-8f508f80-785b-11ea-8f2a-01cebe5dc4cd.png)

Its there, here is the image zoomed 800%

**Comment by ryandemopoulos** [comment]
Thanks for raising this issue @carmellolb.  I think you speak for a lot of folks--including people on the WinUI team itself--by expressing that Background Acrylic is a loved feature and a very important piece of the Fluent story.

@mdtauk and @weltkante both did a great job describing the high level technical details about why this is really hard to do after we decouple from the OS.  As I wrote in the #winui channel of the UWP Community Discord server, my hope is that we can get this addressed asap after WinUI 3.0; I don't know the post-3.0 timeframe because we have more dev engineering to do to settle on how to address it.  But it's a top ask and the community has been very clear that they care about this a whole lot.  I also like the suggestion of us exploring some kind of "faux effect" in the interim; perhaps after //build we can start having deeper discussions on that idea.

@shaheedmalik, I sympathize with your sentiment of not wanting to ship until things are ‘right’; I often feel this way myself and it’s always a balancing act.  We’ve been working with a lot of customers over the last few months, and one thing I’m often reminded of is that WinUI 3.0 will provide a viable path forward for a wide swath of UWP and Win32 apps, and there are many who will be able to use it without relying on the specific effect that background acrylic achieves.  I don’t want to delay unlocking the huge range of features that 3.0 will bring for these customers, so that’s why I continue to push for shipping 3.0 this year and then look to solve the biggest missing pieces (like BG acrylic) as soon as we feasibly can afterward.  I hope that makes sense.

**Comment by mdtauk** [comment]
@ryandemopoulos  If the AcrylicBrush Xaml element can be present, even if it only returns some kind of faux brush fill initially.  Those on 10X, Xbox and Tablet mode devices will look almost identical as there are no overlapping windows.  It is only on Windows 10 devices where the effect will be obviously fake - but if it is only for a 3.0 or 3.1 launch frame, before it is fixed - then at least it wont break code, and will still give the right feel.

If this can be achieved for 3.0 - then that is great.  It will be less jarring than a solid colour fallback.  Also it will allow for the affect to not be disabled on low power devices, or in Tablet mode as currently happens, which is a step up IMO.

**Comment by shaheedmalik** [comment]
> @shaheedmalik, I sympathize with your sentiment of not wanting to ship until things are ‘right’; I often feel this way myself and it’s always a balancing act. We’ve been working with a lot of customers over the last few months, and one thing I’m often reminded of is that WinUI 3.0 will provide a viable path forward for a wide swath of UWP and Win32 apps, and there are many who will be able to use it without relying on the specific effect that background acrylic achieves. I don’t want to delay unlocking the huge range of features that 3.0 will bring for these customers, so that’s why I continue to push for shipping 3.0 this year and then look to solve the biggest missing pieces (like BG acrylic) as soon as we feasibly can afterward. I hope that makes sense.

Then a faux version of the effect needs to be made in the mean time until the correct effect is worked out. The incremental changes of Fluent Design in the past led to the current inconsistency problem that Windows 10 is currently experiencing. If the whole point of WinUI 3 is to fix it, then creating a problem such as this defeating the whole purpose of consistency. Apps with acrylic and without is one of the biggest inconsistencies in Fluent Design.

**Comment by carmellolb** [comment]
> > @shaheedmalik, I sympathize with your sentiment of not wanting to ship until things are ‘right’; I often feel this way myself and it’s always a balancing act. We’ve been working with a lot of customers over the last few months, and one thing I’m often reminded of is that WinUI 3.0 will provide a viable path forward for a wide swath of UWP and Win32 apps, and there are many who will be able to use it without relying on the specific effect that background acrylic achieves. I don’t want to delay unlocking the huge range of features that 3.0 will bring for these customers, so that’s why I continue to push for shipping 3.0 this year and then look to solve the biggest missing pieces (like BG acrylic) as soon as we feasibly can afterward. I hope that makes sense.
> 
> Then a faux version of the effect needs to be made in the mean time until the correct effect is worked out. The incremental changes of Fluent Design in the past led to the current inconsistency problem that Windows 10 is currently experiencing. If the whole point of WinUI 3 is to fix it, then creating a problem such as this defeating the whole purpose of consistency. Apps with acrylic and without is one of the biggest inconsistencies in Fluent Design.

I completely agree. I don’t think WinUI 3 should be released until something is worked out.

**Comment by mdro** [comment]
I actually prefer @mdtauk's static fallback over real transparency. Background acryllic is just distracting when dragging windows, especially with a dark wallpaper and light theme, and led me to disabling transparency completely.

Maybe we should reinterpret coloured backgrounds for non-transient elements (sidebars, title bars, shell etc.) as pure decorations, not as semi-transparency to symbolise a short interruption. This would be more like the titlebar themes for MS Office or Firefox (or IIRC the Vista control center's sidebar). When seen this way, it makes more sense to just use the wallpaper.

It might even be better-looking (and improve performance further?) to make the acryllic backgroud independent of the window position: The wallpaper could either be stretched over the window area or cropped to the top-left corner (or whichever corner makes most sense for the app's resizing behaviour).

**Comment by mdtauk** [comment]
> I actually prefer @mdtauk's static fallback over real transparency. Background acryllic is just distracting when dragging windows, especially with a dark wallpaper and light theme, and led me to disabling transparency completely.
> 
> Maybe we should reinterpret coloured backgrounds for non-transient elements (sidebars, title bars, shell etc.) as pure decorations, not as semi-transparency to symbolise a short interruption. This would be more like the titlebar themes for MS Office or Firefox (or IIRC the Vista control center's sidebar). When seen this way, it makes more sense to just use the wallpaper.
> 
> It might even be better-looking (and improve performance further?) to make the acryllic backgroud independent of the window position: The wallpaper could either be stretched over the window area or cropped to the top-left corner (or whichever corner makes most sense for the app's resizing behaviour).

I would still use the brush surface as a mask so the background stays aligned to the desktop, and remains in place as the window is moved around and resized.  It would just not include any of the windows being visible, just the wallpaper.

This would also have to work with Slideshow backgrounds, as the wallpaper changes, so would the Faux Acrylic.

**Comment by carmellolb** [comment]
I actually disagree with this idea - I think it would be a bit strange to see the desktop wallpaper behind a window when the other windows are behind the window. It would defeat the purpose of depth - and showing layers, which was one of the main purposes of acrylic.

Also, @mdro - how exactly is it distracting? I don't find it distracting at all, in fact I find it supportive of the whole atmosphere, the opposite of distracting.

I think it would look unprofessional, and a bit messed up, because the **material** aspect of Fluent Design is meant to bring aspects of real life, and in real life, something transparent DOES shine to what's right behind it.

I think if faux acrylic becomes a thing, it should appear visually identical to normal acrylic.

**Comment by mdtauk** [comment]
> I think it would be a bit strange to see the desktop wallpaper behind a window when the other windows are behind the window. It would defeat the purpose of depth - and showing layers, which was one of the main purposes of acrylic. It should be identical in appearance to real acrylic.
> 
> I think it would look unprofessional, and a bit messed up if it were to not show what's directly behind it like it does currently, because the **material** aspect of Fluent Design is meant to bring aspects of real life, and in real life, something transparent DOES shine to what's right behind it.
> 
> I think if faux acrylic becomes a thing, it should appear visually identical to normal acrylic.
> 
> Is there no way to show windows behind?
> 
> Why not use semi-transparency like currently?
> 
> How is it distracting? I find it to be the opposite - depth-creating, which Microsoft also said it is when they announced it.
> 
> Changing it to simply a decoration like this would be a step backwards.

At the moment, there are technical hurdles to getting the HostBackdrop AcrylicBrush working in WinUI 3.0

The hope is that those hurdles will be overcome in the future, but not initially for the WinUI 3.0 launch.

What we are discussing is if it is possible to have something other than the standard fallback grey colours until they can get it working.

**Comment by shaheedmalik** [comment]
> I think if this becomes a thing, it should appear visually identical to normal acrylic.
> 
> Something like described before, a faux acrylic but that shines to windows behind and behaves and appears visually identical?
> 
> Is that not possible?
> 
> Why not use semi-transparency like currently?
> 
> How is it distracting? I find it to be the opposite - depth-creating, which Microsoft also said it is when they announced it.
> 
> Changing it to simply a decoration like this would be a step backwards.
> 
> We need acrylic.
Current acrylic isn't possible due to the a technical limitation. We are suggesting a faux acrylic for consistency purposes.
The point is to have something temporary that will not the break the current apps and future apps use of Acrylic until the technical limitation is fixed.

Something is better than nothing.

**Comment by carmellolb** [comment]
> > I think if this becomes a thing, it should appear visually identical to normal acrylic.
> > Something like described before, a faux acrylic but that shines to windows behind and behaves and appears visually identical?
> > Is that not possible?
> > Why not use semi-transparency like currently?
> > How is it distracting? I find it to be the opposite - depth-creating, which Microsoft also said it is when they announced it.
> > Changing it to simply a decoration like this would be a step backwards.
> > We need acrylic.
> 
> Current acrylic isn't possible due to the a technical limitation. We are suggesting a faux acrylic for consistency purposes.
> The point is to have something temporary that will not the break the current apps and future apps use of Acrylic until the technical limitation is fixed.
> 
> Something is better than nothing.

Good point - I just think it would be a bit odd if windows not showing through. I guess this would be the best approach if nothing else is possible though.

**Comment by carmellolb** [comment]
I don't believe this should be used after a solution to the problem is found though, because I'm not sure it would be good for consistency or even just visuals in general long term. Until acrylic can function normally after WinUI 3. @mdro

**Comment by shaheedmalik** [comment]
> I don't believe this should be used after a solution to the problem is found though, because I'm not sure it would be good for consistency or even just visuals in general long term. I think it would be best just until acrylic can function normally after WinUI 3. @mdro

When they fix it in 3.1, the developers of the apps won't have to do anything and the app will just update uniformally.

**Comment by mdtauk** [comment]
> I don't believe this should be used after a solution to the problem is found though, because I'm not sure it would be good for consistency or even just visuals in general long term. I think it would be best just until acrylic can function normally after WinUI 3. @mdro

At the moment Apps running in Tablet Mode fall back and lose their Acrylic.  Low powered devices and VMs also fallback to a solid colour.  If this Faux Acrylic is implemented, it could be used in those scenarios as a performance fallback.

**Comment by carmellolb** [comment]
> > I don't believe this should be used after a solution to the problem is found though, because I'm not sure it would be good for consistency or even just visuals in general long term. I think it would be best just until acrylic can function normally after WinUI 3. @mdro
> 
> At the moment Apps running in Tablet Mode fall back and lose their Acrylic. Low powered devices and VMs also fallback to a solid colour. If this Faux Acrylic is implemented, it could be used in those scenarios as a performance fallback.

Yes, I agree with this. I was referring to that it shouldn't replace normal acrylic once it is fixed.

**Comment by mdro** [comment]
> How is it distracting?

I don't know exactly. I guess the blending of two layers moving relative to each other means that it registers in my brain as changing rather than just moving, which makes it a reason to focus my eyes there.

Anyway, this isn't a huge issue for me, I don't mind the "no transparency" style.
But if you add the fallback anyway, it would be nice if you would take a small bit of extra time and split the transparency effects setting:
1. Interaction effects (reveal etc.) - on/off
2. Transparency (acrylic) - full/reduced (the fallback discussed here and intra-app)/off

(I'd use Interaction on + transparency reduced.)

---

BTW I noticed a few oddities while playing around with the settings a bit more:
- Acrylic on the sidebar of Settings is only active as long as the window has focus. When it looses focus, it instantly switches to the non-transparent gray. So when I have Settings maximized and interact with the tray or notifications, in the corner of my eye I see a large area on the other side of the screen blink to another color. Maybe it should just keep its current acrylic texture when updating the acrylic is suspended due to performance?
- Turning on transparency effects while in powersave mode makes the taskbar background disappear (icons and hover highlights appear on the pure, non-darkened, non-blurred wallpaper).
![Taskbar bug2 2020-04-08 220331](https://user-images.githubusercontent.com/4457627/78832742-13398180-79ec-11ea-95e8-d05c274b38a7.png)
(seen on Surface Pro 6, Update 1909)
- On the same Surface I temporarily saw a dark gray sidebar in Settings as long as transparency was enabled; maybe the background texture didn't get filled and stayed black? The issue disappeared while I played around further .

**Comment by smourier** [comment]
I've just discovered this issue while testing WinUI3.

For me, the main reason why I've been looking at WinUI3 is *precisely* to have an easy access to the backdrop brush and to the acrylic brush (which every customer expect) based on it. 

This is something that's today technically impossible to have w/o using Dcomp (and it's intimacy with DWM, and that's probably the reason why it's difficult to keep in WinUI3 if WinUI3 wants to be independent from Windows). Today, this is accessible using public WinRT API (C#, etc.) though the methods Windows.UI.Composition.Compositor.CreateBackdropBrush / CreateHostBackdropBrush.

If it's not in WinUI, I can (must) continue to use and build a custom acrylic brush manually (in fact, reusing open source WinUI 2 recipes). But since Dcomp is so low level technically, everything is based on it, or not at all. So I can't use WinUI3 at all. Kinda chicken and egg thing.

**Comment by shaheedmalik** [comment]
> @mdtauk and @weltkante both did a great job describing the high level technical details about why this is really hard to do after we decouple from the OS. As I wrote in the #winui channel of the UWP Community Discord server, my hope is that we can get this addressed asap after WinUI 3.0; I don't know the post-3.0 timeframe because we have more dev engineering to do to settle on how to address it. But it's a top ask and the community has been very clear that they care about this a whole lot. I also like the suggestion of us exploring some kind of "faux effect" in the interim; perhaps after //build we can start having deeper discussions on that idea.

Build has come and gone. This needs to be done before WinUI3 releases on a at least faux level for at least consistency purposes. If not, there will be a whole bunch of apps that are inconsistent which will deem WinUI3 a giant waste of time.

**Comment by NicTanghe** [comment]
lol so the one think that makes this framework not useless is being pushed up. Windows sure as fuck don't know how to prioratise stuf ( unless this version allows me to toggle title-bars universally because then all is forgiven. and they made the correct thing the nr 1 priority.)

**Comment by Elttob** [comment]
I'm not really experienced with the internals of Windows UI, but maybe instead of completely omitting background acrylic for the original release, why not make it internally reuse the old acrylic on Windows 10, and fall back to a solid colour on older Windows versions? I think that would be a better compromise until it can be implemented properly.

**Comment by shaheedmalik** [comment]
Since Background Acrylic will not make 3.0, can we have a faux Acrylic in 3.0 in the meantime?

**Comment by Elttob** [comment]
> I'm not really experienced with the internals of Windows UI, but maybe instead of completely omitting background acrylic for the original release, why not make it internally reuse the old acrylic on Windows 10, and fall back to a solid colour on older Windows versions? I think that would be a better compromise until it can be implemented properly.

A placeholder background acrylic brush which just renders the fallback colour would also be acceptable in my mind - that way, the brush can start being used in interfaces right away. Later when the background acrylic brush is properly implemented with proper translucency effects, it would then retroactively apply to those interfaces which used the placeholder, without requiring any code changes.

**Comment by tajetaje** [comment]
Something else that contributes (or so far as I can tell) to this is #1247

**Comment by Poopooracoocoo** [comment]
Also related: #761

**Comment by Poopooracoocoo** [comment]
@mdro 
> it would be nice if you would take a small bit of extra time and split the transparency effects setting:
> * Interaction effects (reveal etc.) - on/off
> * Transparency (acrylic) - full/reduced (the fallback discussed here and intra-app)/off

I pointed out something similar in #816 but they said that it's not their problem. Welcome to the Microsoft world. The most I've seen Microsoft do for itself recently is adding APIs to Windows for Edge (Chromium) and improving WinUI for PowerToys and Terminal. They refuse to update the styles of their other UIs which they obviously still use in Windows. lol

**Comment by shaheedmalik** [comment]
Is Microsoft seriously going to launch WinUI 3.0 without a faux acrylic to cover background acrylic for consistency purposes?

**Comment by NicTanghe** [comment]
If they want it to be independant they just need to do what linux does and have a compositor that always works and runs that handles acrylic ? Blurring.

**Comment by mdtauk** [comment]
> If they want it to be independant they just need to do what linux does and have a compositor that always works and runs that handles acrylic ? Blurring.

Apple seems to be able to handle it's translucency without any significant performance hit - I know Windows is an older code base in many ways - but there is already a Composition Visual Layer, WinUI just needs to be able to call on it and for it to be as performant as can be.

**Comment by NicTanghe** [comment]
> 
> 
> > If they want it to be independant they just need to do what linux does and have a compositor that always works and runs that handles acrylic ? Blurring.
> 
> Apple seems to be able to handle it's translucency without any significant performance hit - I know Windows is an older code base in many ways - but there is already a Composition Visual Layer, WinUI just needs to be able to call on it and for it to be as performant as can be.

Here`s to hoping upstream wil not point fingers to other people asking them to fix it in a shitty way because they can make a shitty workaround.

**Comment by mdtauk** [comment]
I wonder if going back to Windows 7 - WinUI/Reunion could provide a driver, which provides the Rendered layer needed to handle Acrylic.  WinUI apps on deployment and new OS versions could include the driver which provides the base layer render visual, via the Store.

Windows 7 and Vista before it must have had the capability for the Aero Glass rendering.  Windows 8/8.1 had the Visual Composition Layer.  So perhaps this is more of a Reunion ask, but if the driver handled the OS dependent stuff, the rest could all remain User level right?

**Comment by SFM61319** [comment]
Yea so I found [this concept of fluent ui](https://www.reddit.com/r/Windows10/comments/j5crgf/windows_10_fluent_design_with_blur_effects_concept/?utm_medium=android_app&utm_source=share) on Reddit and thought looked clean and "fluent".

**Comment by Jaiganeshkumaran** [comment]
Fluent was originally created due to the lack of transparency complains from users. But now you are removing it entirely.

**Comment by mdtauk** [comment]
> Fluent was originally created due to the lack of transparency complains from users. But now you are removing it entirely.

Microsoft has said they need more time to make Acrylic work, when WinUI is no longer part of the OS, but running separately

**Comment by Jaiganeshkumaran** [comment]
I don't care about acrylic anyway, however, hope they will not remove it completely.

**Comment by Jaiganeshkumaran** [comment]
I don't understand why WinUI 3 can't support Acrylic. WinUI 3 doesn't need to do the work, let it handle it the compositor on the supported platform and no need to make it open source.

**Comment by reinux** [comment]
Here's some history.

Performance is not the reason. It never was. It was almost certainly a PR decision.

Acrylic has been around since 2007. Back then, it was called Aero Glass, and everyone hated it because it supposedly decimated battery life because it required the GPU.

What people didn't realize (or didn't want to realize, because it was sensational) was that doing window compositing on the GPU was not only faster but also more efficient, because that's specifically the kind of task that the GPU is wired for. Once the drivers came in (which didn't always happen because [OEMs could just blame Vista for all their woes](https://www.theregister.com/2008/07/22/sandisk_ssd_vista_beef/)), Vista battery life was generally as good as XP, with or without glass and with or without Aero (the GPU compositing component).

Part of the misconception came from the claim that Vista "required" a modern GPU with DX9 and Shader 2.0 support to run, which the media walked back when it became clear that Glass could be turned off. But since people generally equate newer hardware with faster hardware, people didn't really understand that it wasn't really a computing power requirement so much as it was a hardware _feature_ requirement.

There were a few articles that pointed actually bothered to benchmark or at least test battery consumption with and without Glass: [1](https://www.pcworld.com/article/131885/article.html) [2](http://brandonlive.com/2007/05/04/vista-battery-life-error-in-cnet-article/) [3](https://www.infoworld.com/article/2640522/the-vista-aero-vs--battery-life-myth.html), to list a few, and found no more than 1-4% difference in battery life with Glass. I also found no measurable difference on my Toshiba Portege M400, which sported 512MB of RAM and something like 8MB of GPU memory. All the other reports were entirely anecdotal and included no actual measurements.

Microsoft, being caught in a weird spot between fact and a blogosphere that simply wouldn't have it, had forced itself to slowly phase out Glass in 7 and then fully in 10, even as iOS and others began adopting similar effects.

Other than maybe something like a Raspberry Pi, there's simply no hardware in 2020 that couldn't efficiently handle an effect that my 2005 potato laptop could handle so effortlessly.

**Comment by reinux** [comment]
That said, I want to pitch a compromise: Could Acrylic just paint part of the desktop wallpaper instead of all the windows behind it? Or maybe have that as an option?

Some old Linux terminals in the early 2000s did a thing where it would just paint the desktop wallpaper image at reduced contrast, and it would move the source rect based on where the window was relative to the wallpaper.

This has a few advantages:

- Wallpapers choice can be made more relevant
- It can look more consistent and less busy/cluttered than if it's rendering user windows behind it
- The effect can be cached, and there's guaranteed to be no animation behind it, which would put any concerns about performance/battery life to rest for good (it shouldn't cost more than it would just moving your window around, except maybe if you're using Remote Desktop or similar -- in which case Glass/Acrylic is disabled automatically anyway)

I made a mockup a couple years ago in WPF:

![ezgif-7-ad2b25eee91b](https://user-images.githubusercontent.com/5127297/100915923-e7db3980-3489-11eb-9c72-9150b9059149.gif)

**Comment by mdtauk** [comment]
@reinux  I proposed using the Desktop Wallpaper as a fallback option, and still think it would be a good middle ground, not just temporary whilst Acrylic is made to work with WinUI, but also for Inactive Windows, Tablet Mode, and other performance restricted devices.

AeroGlass was less GPU intensive as it was a blur over the back rendering, with a colour fill and PNG overlays to create the tinting and shimmers.  

Acrylic involves the stronger Blurring with Saturation Boost layers, a Noise layer, as well as Luminosity and Tint layers, all computed in real time.

**Comment by reinux** [comment]
My bad, I should have read the prior discussion more carefully.

**Comment by mdtauk** [comment]
> My bad, I should have read the prior discussion more carefully.

Your mockup is a great illustration!  Thank you for posting it.

**Comment by DiegoMagdaleno** [comment]
So is Acrylic a planed thing? Until it is independent from the Operating System itself and depends on WinUI?

**Comment by chingucoding** [comment]
Yes Acrylic is a planed thing, however it will not arrive with WinUI 3.0. According to the [WinUI 3 roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md) will arrive with a later WinUI 3 version.

**Comment by axcadsd** [comment]
I would also prefer the Desktop Wallpaper faux acrylic effect. Can windows team maybe consider adding a global setting through which we might be able to use the faux effect globally for all apps that support acrylic? Seeing the background applications through acrylic effect is distrating and sometimes reduces readability...

**Comment by legistek** [comment]
Sorry for asking a dumb question here, but I am just starting to experiment with the newest WinUI 3 preview and I tried the AcrylicBrush. It is always just going straight to the fallback color and never giving me an effect. This is expected behavior at this point correct? And it sounds like that's not going to change before release?

**Comment by legistek** [comment]
Also it's been said, including here, that Win UI 3 is somehow de-coupled from the OS, but I've been looking at the source code in this repo and I don't see how that is the case at all. The C++ code is still extensively relying on the underlying WinRT APIs which only ship _with Windows_. In fact it looks like mostly wrappers for WinRT classes that allow the latter to be used in Win32 apps. 

Therefore I don't understand when people refer to re-implementing something like the AcryllicBrush for Win UI 3; it's not like they're re-implementing Button or Grid or anything else low level. 

Am I missing something here?

**Comment by chingucoding** [comment]
> Also it's been said, including here, that Win UI 3 is somehow de-coupled from the OS, but I've been looking at the source code in this repo and I don't see how that is the case at all. The C++ code is still extensively relying on the underlying WinRT APIs which only ship with Windows. In fact it looks like mostly wrappers for WinRT classes that allow the latter to be used in Win32 apps.
>
> Therefore I don't understand when people refer to re-implementing something like the AcryllicBrush for Win UI 3; it's not like they're re-implementing Button or Grid or anything else low level.
>
> Am I missing something here?

This is because the source code of this project is WinUI 2.x as of right now, not WinUI 3. Once WinUI 3 goes open source, it's source code will be merged into this repository, until then, all you can see is the WinUI 2.x source code which still relies on UWP XAML Controls and WinRT APIs.

**Comment by chingucoding** [comment]
> Sorry for asking a dumb question here, but I am just starting to experiment with the newest WinUI 3 preview and I tried the AcrylicBrush. It is always just going straight to the fallback color and never giving me an effect. This is expected behavior at this point correct? And it sounds like that's not going to change before release?

Yes this is more or less expected and will be fixed with a WinUI 3.x release according to the roadmap.

**Comment by legistek** [comment]
> This is because the source code of this project is WinUI 2.x as of right now, not WinUI 3. Once WinUI 3 goes open source, it's source code will be merged into this repository, until then, all you can see is the WinUI 2.x source code which still relies on UWP XAML Controls and WinRT APIs.

Ohhh, ok that explains a lot. I got thrown off by the fact that the repo has so many recent commits but I guess they're still actively maintaining 2.x too.

Well that's exciting. I wonder if it'll be feasible to (unofficially) port to other platforms like web and mobile then.

**Comment by chingucoding** [comment]
> Ohhh, ok that explains a lot, thanks! I got thrown off by the fact that the repo has so many recent commits but I guess they're still actively maintaining 2.x too.

Yes WinUI 2.x is being actively developed with new features and style updates being added. This will also be the case for multiple months even while WinUI 3 has been released.

> Well that's exciting. I wonder if it'll be feasible to (unofficially) port to other platforms like web and mobile then.

I have excellent news for you! This is exactly what the Uno Platform team is doing! You can check their website out https://platform.uno/ or view their source code here: https://github.com/unoplatform/uno

**Comment by legistek** [comment]
> I have excellent news for you! This is exactly what the Uno Platform team is doing! You can check their website out https://platform.uno/ or view their source code here: https://github.com/unoplatform/uno

Oh yes I've had 3 eyes on Uno for awhile now. :) They're recreating the XAML syntax though, and not necessarily actually porting Microsoft's actual open source code right?

**Comment by chingucoding** [comment]
> Oh yes I've had 3 eyes on Uno for awhile now. :) They're recreating the XAML syntax though, and not necessarily actually porting Microsoft's actual open source code right?

It's more or less a recreation and no port, yes. Though for WinUI 2 controls iirc, they closely use look at the WinUI 2 code base and copy the existing code instead of reinventing the wheel. Maybe @martinzikmund can weigh in on this as he is a very active contributor to the Uno platform and helped porting WinUI 2 controls to Uno!

**Comment by MartinZikmund** [comment]
Hey, yes, that is true 🙂 . While the majority of the codebase is a custom implementation, we are trying to port the open-source parts of WinUI line-by-line (by translating to C#), to achieve exact functional parity. When the rest of WinUI gets open-source, we will definitely revisit the base controls as well 👍 . The ultimate goal is for you to not have to care about the platform-specifics and just write your Windows app to have it run everywhere. There is still a lot of work to do, but we are not planning to slow down 🚀

**Comment by legistek** [comment]
> Hey, yes, that is true 🙂 . While the majority of the codebase is a custom implementation, we are trying to port the open-source parts of WinUI line-by-line (by translating to C#), to achieve exact functional parity. When the rest of WinUI gets open-source, we will definitely revisit the base controls as well 👍 . The ultimate goal is for you to not have to care about the platform-specifics and just write your Windows app to have it run everywhere. There is still a lot of work to do, but we are not planning to slow down 🚀

I think the work you guys are doing is extremely impressive, don't get me wrong. You're doing what Microsoft really should have been doing on their own for 10 years now. I think it's a great idea to import as much FOSS code from Microsoft as you can. No hard feelings against them for not doing it, it's their prerogative how they spend their resources, but if they don't someone should!

**Comment by AnderssonPeter** [comment]
I was soo hyped when winui3 0.5 launched, but then saw that background acrylic was planned as post 0.8.
How sad now I'm not hyped at all any more, I hope they manage to get it working for 0.8.

**Comment by ryandemopoulos** [comment]
Hello @AnderssonPeter, background acrylic has been planned as a post 0.8 thing for quite a while, but perhaps you haven't followed (or weren't aware) of our public roadmap which is here: https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md

We added in-app acrylic for 0.5 -- that should be working -- and we're still sorting out the timeframe & technical solution for post-0.8.  It's being evaluated alongside a lot of other vital work that needs to be done, including unpackaged deployment support, XAML Islands, multi-window support (using separate threads), and more.  Once we have the plan nailed down we'll update the public roadmap.

**Comment by legistek** [comment]
> We added in-app acrylic for 0.5 -- that should be working

As far as I can tell it still just equivalent to a SolidColorBrush.  Am I not using it right?

**Comment by mgoyalitm** [comment]
What about using IHostbackdropProvider interface in dotnet 5 and use dll import somehow and wire it up with windows at least, probably a nuget package WinUI.Lasthope.Windows?

Or Just do it in your own way.

**Comment by johngagefaulkner** [comment]
> Hello @AnderssonPeter, background acrylic has been planned as a post 0.8 thing for quite a while, but perhaps you haven't followed (or weren't aware) of our public roadmap which is here: https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md
> 
> We added in-app acrylic for 0.5 -- that should be working -- and we're still sorting out the timeframe & technical solution for background acrylic post-0.8. It's being evaluated alongside a lot of other vital work that needs to be done, including unpackaged deployment support, XAML Islands, multi-window support (using separate threads), and more. Once we have the plan nailed down we'll update the public roadmap.

Slightly off-topic here, but, I'm having a hard time finding any mention of XAML Designer support on any roadmap. It's definitely listed under "Known Issues" with the patch notes of each release but I'm curious if you can speak to its priority amongst the myriads of other tasks the team has? Thanks in advance!

**Comment by chrisgull** [comment]
Still no AcrylicBrush. This is sorely needed for us who use transparency extensively in our UI.
I'm just now porting my Xamarin apps (which have acrylic or similar working on UWP/iOS/Android) to MAUI, and the lack of support for AcrylicBrush in WinUI 3 will affect my user experience in a negative way.
Please get this done soon.

**Comment by sharpninja** [comment]
It seems to me that the OS code that does Acrylics could be exposed as an API which would return a primitive that would be applied to a FrameworkElement as the BackgroundBrush.  If the primitive moves, the OS would be notified so it can rerender without any extra work by the App.  The API could be public or cryptographically tied to the public thumbprint of a Store app.

**Comment by Foxlider** [comment]
I've been searching everywhere for Acrylic support for Windows10 on WinUI3 but still no luck sadly...

**Comment by dajbych** [comment]
According to [Windows App SDK roadmap](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md) Mica & Background Acrylic should be already available since the first half of 2022. Planned features on the [product dashboard](https://portal.productboard.com/winappsdk/1-windows-app-sdk/tabs/2-planned) includes it in the version 1.1 in Q2 2022. [Microsoft documentation](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#version-113) is mentioning that the current version is 1.1.3. I'm confused. Upgrading from UWP to the Windows App SDK is difficult to plan.

**Comment by KaibaiToday** [comment]
I agree. Why did acrylics brush discontinue the "BackgroundSource" property?

**Comment by lyahdav** [comment]
Per https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#winui-3 in WinAppSDK 1.1:
> Mica and Background Acrylic are now available for WinUI 3 applications

If someone can verify this works, can we close this issue?

**Comment by dajbych** [comment]
I does not work (at least on Windows 10). _AcrylicBackgroundFillColorDefaultBrush_ does not make transparent window. When I tried it by code, the _AcrylicBrush_'s property _BackgroundSource_ is not defined. The _HostBackdrop_ class cannod be found in WinAppSDK 1.1 documentation. Apply Mica or Acrylic documentation is mentioning Windows 11 only. Transparent windows are [supported since Windows Vista](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/graphics-multimedia/extend-glass-frame-into-a-wpf-application?view=netframeworkdesktop-4.8). And I want to upgrade my Windows 10 compatible UWP app to Windows App SDK only (and keep Windows 10 compatibility).

**Comment by lyahdav** [comment]
Someone might want to investigate getting the HWND for the WinUI 3 window and using win32 APIs to make it transparent.

**Comment by codendone** [comment]
Window transparency is not currently supported in WinAppSDK. Adjusting HWND properties will not be able to enable transparency.

Background Acrylic is supported now in WinAppSDK 1.1, as @lyahdav linked to above. However, it works differently in WinAppSDK than it does in WinUI 2 ([docs and sample code](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#winui-3)), and it is only supported on Windows 11 in WinAppSDK 1.1. The upcoming WinAppSDK 1.2 release adds support for Background Acrylic (via `DesktopAcrylicController`) on Windows 10; this is available for testing in 1.2-preview2 (see #7112).

**Comment by lyahdav** [comment]
@codendone I was able to get window transparency working in WinAppSDK, at least for our needs. We need the window to not be rectangular and have parts of it transparent. I was able to get this working with the code below. This was on 
 WinAppSDK 1.0.4. Are you saying that this functionality breaks in newer versions of WinAppSDK?

Similarly I was able to get a fully transparent window following the approach [here](https://stackoverflow.com/a/3970218/62), but then all elements inside the window are transparent, which isn't what we want.

```
HWND GetHWNDFromWindow(MUX::Window const &window) {
  HWND nativeWindow{nullptr};
  winrt::check_hresult(window.as<IWindowNative>()->get_WindowHandle(&nativeWindow));
  return nativeWindow;
}

...

  const auto hwnd = GetHWNDFromWindow(window);
  SetWindowLong(
      hwnd,
      GWL_STYLE,
      GetWindowLong(hwnd, GWL_STYLE) & ~WS_CAPTION & ~WS_THICKFRAME); // remove title bar and sizing border

  RECT windowRect;

  auto hOrgRgn = CreateRectRgn(0, 0, 0, 0);
  GetWindowRgn(hwnd, hOrgRgn);
  GetWindowRect(hwnd, &windowRect);

  auto hRoundRectRgn = CreateRoundRectRgn(
      0,
      0,
      windowRect.right - windowRect.left,
      windowRect.bottom - windowRect.top,
      windowRect.right - windowRect.left,
      windowRect.bottom - windowRect.top);
  winrt::check_hresult(SetWindowRgn(hwnd, hRoundRectRgn, TRUE));
```

Here's a screenshot:
![image](https://user-images.githubusercontent.com/359157/200708440-7aa9c7b5-5afe-430c-b9a0-50bfb98f8bd0.png)

**Comment by lyahdav** [comment]
Actually I just tried with WinAppSDK 1.2.220930.4 and the code above still works.

**Comment by chrisgull** [comment]
> Actually I just tried with WinAppSDK 1.2.220930.4 and the code above still works.

The topic here is Acrylic (blurred), not transparency.

**Comment by castorix** [comment]
> SetWindowRgn

Transparency with Regions has always worked, but it is an old method, which does not use anti-aliasing like Layered windows.

**Comment by bogdan-patraucean** [comment]
This is now part of WinUi 3. Maybe the issue could be closed?

**Comment by codendone** [comment]
Desktop Acrylic can now be easily applied to the window WinAppSDK 1.3 via [DesktopAcrylicBackdrop](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.desktopacrylicbackdrop?view=windows-app-sdk-1.3).

**Comment by Crypties** [comment]
Having read this thread one has to wonder: is this 2020 thread the genesis of Mica, inspired by a user coming with this idea, calling it then "Faux Acrylic"? @mdtauk I think you may have inspired Mica.

**Comment by shenxianovo** [comment]
I just finished reading all the comments to these issue, but I still have some questions.
In Microsoft's official documentation about [Acrylic](https://learn.microsoft.com/en-us/windows/apps/design/style/acrylic), there is a property `BackgroundSource` in the "Custom Acrylic Brush" section of the code. However, this property can't be found in class `AcrylicBrush ` at all.

``` xaml
<ResourceDictionary.ThemeDictionaries>
    <ResourceDictionary x:Key="Default">
        <AcrylicBrush x:Key="MyAcrylicBrush"
            BackgroundSource="HostBackdrop"/> <!-- HERE -->
    </ResourceDictionary>
</ResourceDictionary.ThemeDictionaries>
```

Does this mean that the property will be implemented later? Or does it indicate that all development related to Acrylic has stopped for now?

**Comment by weltkante** [comment]
@shenxianovo thats WinUI2, for WinUI3 you use different API, see this part on the same documentation page you linked: https://learn.microsoft.com/en-us/windows/apps/design/style/acrylic#apply-acrylic-in-your-app
